### PR TITLE
Reimplement hipMemGetInfo

### PR DESCRIPTION
**Pull request by satyanveshd**
Addresses SWDEV-136570. hipMemGetInfo changed to compute free memory based on information from kfd instead of relying on hc::am_tracker.

### CODE REVIEW

File: src/hip_memory.cpp [hunk 1]
```diff
@@ -1969,14 +1970,27 @@ hipError_t hipMemGetInfo(size_t* free, size_t* total) {
...
+		
+		size_t deviceMemSize;
+		std::string fileName = std::string("/sys/class/kfd/kfd/topology/nodes/") + std::to_string(internal_node_id) + std::string("/mem_banks/0/used_memory");  
+                
+		std::ifstream file;
+		file.open(fileName);
+		if(!file) return ihipLogStatus(hipErrorFileNotFound);
+		
+		file >> deviceMemSize;    
+		*free = device->_props.totalGlobalMem - deviceMemSize;
```
Comment (SarbojitAMD) [inline src/hip_memory.cpp]: I feel it is good to have a check on read data i.e. deviceMemSize. There is some existing function e.g. isdigit() can be used to make sure read value is indeed a number.

File: src/hip_memory.cpp [hunk 2]
```diff
@@ -1969,14 +1970,27 @@ hipError_t hipMemGetInfo(size_t* free, size_t* total) {
...
         if (free) {
-            // TODO - replace with kernel-level for reporting free memory:
-            size_t deviceMemSize, hostMemSize, userMemSize;
-            hc::am_memtracker_sizeinfo(device->_acc, &deviceMemSize, &hostMemSize, &userMemSize);
-
-            *free = device->_props.totalGlobalMem - deviceMemSize;
-
-            // Deduct the amount of memory from the free memory reported from the system
-            if (HIP_HIDDEN_FREE_MEM) *free -= (size_t)HIP_HIDDEN_FREE_MEM * 1024 * 1024;
+            hsa_agent_t* agent = static_cast<hsa_agent_t*>(device->_acc.get_hsa_agent());
```
Comment (SarbojitAMD) [inline src/hip_memory.cpp]: Looks like implementation assumes that 'device' will be not null if 'context' exists, can you please confirm that?

File: src/hip_memory.cpp (hunk 1 above)
Comment (satyanveshd) [reply src/hip_memory.cpp]: used strtol() as the usedmemory can be large,  to check the value read is a number. Updated the PR.

File: src/hip_memory.cpp (hunk 2 above)
Comment (satyanveshd) [reply src/hip_memory.cpp]: Yes, as we cannot create a context without having an associated device.

File: src/hip_memory.cpp [hunk 3]
```diff
@@ -1969,14 +1970,34 @@ hipError_t hipMemGetInfo(size_t* free, size_t* total) {
...
+            if (!agent){
+		return ihipLogStatus(hipErrorInvalidResourceHandle);
+            } else {
+		uint32_t internal_node_id;
+		hsa_status_t err = hsa_agent_get_info(*agent, (hsa_agent_info_t) HSA_AMD_AGENT_INFO_DRIVER_NODE_ID, &internal_node_id);
```
Comment (mangupta) [inline src/hip_memory.cpp]: We don't need to query for nodeID everytime. So a better idea would be to store this information in ihipDevice_t and initialize the value in https://github.com/ROCm-Developer-Tools/HIP/blob/master/src/hip_hcc.cpp#L504.

File: src/hip_memory.cpp [hunk 4]
```diff
@@ -1969,14 +1970,34 @@ hipError_t hipMemGetInfo(size_t* free, size_t* total) {
...
+		std::string fileName = std::string("/sys/class/kfd/kfd/topology/nodes/") + std::to_string(internal_node_id) + std::string("/mem_banks/0/used_memory");  
+                
+		std::ifstream file;
+		file.open(fileName);
+		if (!file) return ihipLogStatus(hipErrorFileNotFound);
+	
+                std::string deviceSize;	
+		size_t deviceMemSize;
+
+		file >> deviceSize;
```
Comment (mangupta) [inline src/hip_memory.cpp]: close file immediately after this instead having to deal with it in both the if() as well as else() cases.

File: src/hip_memory.cpp [hunk 5]
```diff
@@ -1969,14 +1970,34 @@ hipError_t hipMemGetInfo(size_t* free, size_t* total) {
...
+                if ((deviceMemSize=strtol(deviceSize.c_str(),NULL,10))){
+		    *free = device->_props.totalGlobalMem - deviceMemSize;
+		} else {
+                    file.close();
+ 		    return ihipLogStatus(hipErrorInvalidValue);
+		}
+		file.close();                 
+            	// Deduct the amount of memory from the free memory reported from the system
+		if (HIP_HIDDEN_FREE_MEM) *free -= (size_t)HIP_HIDDEN_FREE_MEM * 1024 * 1024;
```
Comment (mangupta) [inline src/hip_memory.cpp]: This can be moved into the if() case above.

File: src/hip_hcc_internal.h [hunk 6]
```diff
@@ -802,6 +802,9 @@ class ihipDevice_t {
 
     // TODO - report this through device properties, base on HCC API call.
     int _isLargeBar;
+   
+    // Node id reported by kfd for this device
+    uint32_t internal_node_id;
```
Comment (mangupta) [inline src/hip_hcc_internal.h]: Rename to _node_id or _driver_node_id

Comment (satyanveshd) [reply src/hip_hcc_internal.h]: renamed to _driver_node_id. updated the PR.

**Pull request merged**